monai/engines/workflow.py

```diff
@@ -119,44 +119,53 @@ def set_sampler_epoch(engine: Engine):
+            self._register_post_transforms(post_transform)
         if key_metric is not None:
-
-            if not isinstance(key_metric, dict):
-                raise TypeError(f"key_metric must be None or a dict but is {type(key_metric).__name__}.")
-            self.state.key_metric_name = list(key_metric.keys())[0]
-            metrics = key_metric
-            if additional_metrics is not None and len(additional_metrics) > 0:
-                if not isinstance(additional_metrics, dict):
-                    raise TypeError(
-                        f"additional_metrics must be None or a dict but is {type(additional_metrics).__name__}."
-                    )
-                metrics.update(additional_metrics)
-            for name, metric in metrics.items():
-                metric.attach(self, name)
-
-            @self.on(Events.EPOCH_COMPLETED)
-            def _compare_metrics(engine: Engine) -> None:
-                if engine.state.key_metric_name is not None:
-                    current_val_metric = engine.state.metrics[engine.state.key_metric_name]
-                    if current_val_metric > engine.state.best_metric:
-                        self.logger.info(f"Got new best metric of {engine.state.key_metric_name}: {current_val_metric}")
-                        engine.state.best_metric = current_val_metric
-                        engine.state.best_metric_epoch = engine.state.epoch
-
+            self._register_metrics(key_metric, additional_metrics)
         if handlers is not None:
-            handlers_ = ensure_tuple(handlers)
-            for handler in handlers_:
-                handler.attach(self)
-        self.amp = amp
+            self._register_handlers(handlers)
+
+    def _register_additional_events(self):
+        pass
+
+    def _register_post_transforms(self, posttrans):
+        @self.on(Events.ITERATION_COMPLETED)
+        def run_post_transform(engine: Engine) -> None:
+            if posttrans is None:
+                raise AssertionError
+            engine.state.output = apply_transform(posttrans, engine.state.output)
+
+    def _register_metrics(self, k_metric, add_metrics):
+        if not isinstance(k_metric, dict):
+            raise TypeError(f"key_metric must be None or a dict but is {type(key_metric).__name__}.")
+        self.state.key_metric_name = list(k_metric.keys())[0]
+        metrics = k_metric
+        if add_metrics is not None and len(add_metrics) > 0:
+            if not isinstance(add_metrics, dict):
+                raise TypeError(
+                    f"additional_metrics must be None or a dict but is {type(additional_metrics).__name__}."
+                )
+            metrics.update(add_metrics)
+        for name, metric in metrics.items():
+            metric.attach(self, name)
+
+    def _register_handlers(self, handlers):
+        handlers_ = ensure_tuple(handlers)
+        for handler in handlers_:
+            handler.attach(self)
+
+        @self.on(Events.EPOCH_COMPLETED)
```

Comment:
this could be a generic event handler for best metric/loss

Reply:
Sorry for my mistake, this callback should be in the `_register_metrics()` function instead of handlers, because it depends on the `state.best_metric` and `state.best_metric_epoch` which only exist in MONAI workflows, and should only work when metrics registered.
I updated this part in the latest commit.

Thanks.